### DISCUSSION TITLE
Edge Functions to cut down cold starts. Serverless as it should be

### DISCUSSION THREAD

**Discussion by fungilation** [nhost/nhost#1970] [Ideas]
As can be seen with new functions metrics, real world from my app: https://twitter.com/nhost/status/1662146263452622848

Cold starts are 2+ seconds. The only main drawback to serverless generally I'd say. This can largely be cut down with V8 isolates/edge runtimes, like cloudflare/vercel's and Deno.

Bonuses:
- pricing and charge on CPU time, not on wall clock (IO await)
- potentially unlimited run time duration, as long as response returned within X (30s on Vercel wall clock)
- global CDN from which the lightweight edge functions can run. But this is the least of the benefits to me, as most edge functions I see have dependency in querying our data layer in hasura/postgres which is for most cases centralized to 1 region. So there should be config to define per function, which stateless functions that can run on cdn (closest to client), and which to run close to db region

Downsides:
- V8 isolate has limited api, not full Node's. But I think large majority of functions use cases don't need it (but then Deno comes and blog about almost supporting the entire set of Node.js modules: https://deno.com/blog/node-builtins-on-deploy)
- 128mb memory hard cap. Fine tradeoff in most cases, alternative is good old Node.js

My take is Edge function's runtime has characteristics that makes Edge functions the final form of serverless: ephemeral, lightweight, to the second compute based pricing so don't have to think about it, the primary workhorse for most IO-bound BE processes. Serverless as it should be.

**Comment by dbarrosop** [comment]
Thanks, this is very useful. We need to revisit our current functions' solution so we will keep in mind this when we do so.

**Comment by eqoram** [comment]
+1 for cloudflare workers! This project might be of interest for it: https://github.com/hisorange/opwork

**Comment by dionjwa** [comment]
I would love support for deno.deploy edge functions (would also be good with cloudflare workers). The features and ergonomics mean I am going to use it anyway for my edge functions, and just live with the "database in a far region" for now.
